### PR TITLE
'/imsvc' and '/external:I' flags support for MSVC

### PR DESCRIPTION
**Pull request by poppolopoppo**
# Description:

This PR tries to address the issues described in #764, which yield to distribution not working when those flags are used.
The patch is small, but it seems to be all that was needed ?

# Checklist:

The pull request:
- [X] **Is created against the Dev branch**
- [X] **Is self-contained**
- [X] **Compiles on Windows, OSX and Linux**
- [ ] **Has accompanying tests**
- [X] **Passes existing tests**
- [ ] **Keeps Windows, OSX and Linux at parity**
- [X] **Follows the code style**
- [ ] **Includes documentation**

### CODE REVIEW

File: Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp [hunk 1]
```diff
@@ -1748,10 +1748,13 @@ bool ObjectNode::BuildArgs( const Job * job, Args & fullArgs, Pass pass, bool us
...
                 // limit of cl.exe)
-                if ( ( job->IsLocal() == false ) && IsStartOfCompilerArg_MSVC( token, "I" ) )
+                if ( ( job->IsLocal() == false ) && (
+                        IsStartOfCompilerArg_MSVC( token, "I" ) ||
+                        IsStartOfCompilerArg_MSVC( token, "imsvc" ) ||
+                        IsStartOfCompilerArg_MSVC( token, "external:I" ) ) )
                 {
                     // Get include path part
-                    const char * start = token.Get() + 2; // Skip /I or -I
+                    const char * start = token.Get() + 2; // Skip /I or -I or /imsvc or -imsvc or /external:I or -external:I
```
Comment (ffulin) [inline Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp:1757]: This still seems to add +2 to the token.Get(), which doesn't seem correct for "imsvc" and "external:I"

Comment (poppolopoppo) [reply Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp:1757]: Yes you are absolutely right,
I'll update this patch soon after additional testing.

**Comment by ffulin** [comment]
It seems /external:I can't currently be supported with distributed compilation. See: #764 for more details.

**Comment by ffulin** [comment]
For reference, I changed this code to fix a couple of issues. I have this shelved should it become possible to use (if the MSVC issues are resolved):
```C++
                if ( job->IsLocal() == false )
                {
                    // Check for various styles of include token
                    static const char * const includeTokens[] =
                    {
                        "I",
                        "imsvc",
                        "external:I",
                    };
                    const char * start = nullptr;
                    for ( const char* includeToken : includeTokens )
                    {
                        if ( IsStartOfCompilerArg_MSVC( token, includeToken ) )
                        {
                            start = token.Get() + ( AString::StrLen( includeToken ) + 1 ); // +1 for - or /
                            break;
                        }
                    }
                    if ( start )
                    {
                        // Get include path part, removing quotes if present
                        if ( *start == '"' )
                        {
                            ++start;
                        }
                        const char * end = token.GetEnd();
                        if ( end[ -1 ] == '"' )
                        {
                            --end;
                        }
                        AStackString<> includePath( start, end );
                        const bool isFullPath = PathUtils::IsFullPath( includePath );

                        // Replace relative paths and leave full paths alone
                        if ( isFullPath == false )
                        {
                            // Remove relative include
                            for ( const char* includeToken : includeTokens )
                            {
                                if ( StripTokenWithArg_MSVC( includeToken, token, i ) )
                                {
                                    break;
                                }
                            }

                            // Add full path include
                            fullArgs.Append( token.Get(), (size_t)( start - token.Get() ) );
                            fullArgs += job->GetRemoteSourceRoot();
                            fullArgs += '\\';
                            fullArgs += includePath;
                            fullArgs.Append( end, (size_t)( token.GetEnd() - end ) );
                            fullArgs.AddDelimiter();

                            continue; // Include path has been replaced
                        }
                    }
                }
```

**Comment by ffulin** [comment]
The issue i originally raised in the Developer Community was incorrectly closed as a duplicate, but I re-reported the issue, split in two:
* https://developercommunity2.visualstudio.com/t/external:I-requires-directory-to-exist/1340525
* https://developercommunity2.visualstudio.com/t/experimental:I-incorrectly-ignores-dire/1306689

It seems those issues have now been closed as 'Fixed In: Visual Studio 2019 version 16.10'. Pending verification of those fixes, we can hopefully resolve this soon.

**Comment by ffulin** [comment]
As with the related issue (https://github.com/fastbuild/fastbuild/issues/764) I think this PR slipped through the cracks. FASTBuild has been working with '/imsvc' and '/external:I' for some time now (since around v1.08 or so), assuming a recent enough version of the MSVC compiler to avoid the issues it had originally.

I'll close this as there are no know outstanding issues. We can of course raise new issues for any problems that are found.

**Pull request closed without merging**